src/python/gudhi/wasserstein/quantization.py

```diff
@@ -0,0 +1,189 @@
+    else:
+        a = np.argmin(M[:-1, :-1], axis=1)
+
+    cells = [X[a == j] for j in range(k)]
+
+    if withdiag:
+        cells.append(X[a == k])  # this is the (k+1)-th centroid corresponding to the diagonal
+
+    return cells
+
+
+def _from_batch(pdiagset, batches_indices):
+    """
+    Given a pdiagset and a list of indices, compute the "empirical expected persistence diagram"
+    (in practice, concatenates) on the corresponding batch of diagrams.
+    Note that we may encounter empty diagrams, hence the specific check .ndim==2.
+        If the resulting list is empty (all diagrams are empty), returns empty array.
+    """
+    list_of_non_empty_diags = [pdiagset[i] for i in batches_indices if pdiagset[i].ndim==2]
+    if list_of_non_empty_diags: # There are some non-empty diagrams
+        X_batch = np.concatenate(list_of_non_empty_diags)
+        return X_batch
+    else:
+        return np.array([])
+
+
+def _init_c(pdiagset, k, internal_p=2):
+    """
+    A naive heuristic to initialize a codebook: we take the k points with largest distances to the diagonal
+        in the first diagram of the list.
+
+    Parameters
+    ----------
+    :param list_diags: The list of diagrams we want to quantize. Assume it is not empty (checked by quantization).
+    :param k: number of centroids in the codebook.
+    :param internal_p: Internal ground metric. Should not play a role specific role.
+
+    :returns: an initial codebook c0 with `k` entries.
+    -------
+
+    """
+    dgm = pdiagset[0]
+    w = _utils._dist_to_diag(dgm, internal_p)
+    s = np.argsort(w)
+    c0 = dgm[s[-k:]]  # Elements in dgm with larger persistence are last
+    return c0
+
+
+#####################
+###  Main method  ###
+#####################
+
+def quantization(pdiagset, k=2, init=None, batch_size=1, order=2., internal_p=2.):
+    """
+    This quantization algorithm takes a list of diagrams ``pdiagset``, an integer ``k``
+    (or an initial codebook guess ``c0``)
+    and returns a codebook of ``k`` centroids that aims at minimizing the distance between the codebook and the
+    expected persistence diagram underneath the diagrams in ``pdiagset``.
+    The codebook is iteratively updated by going through (batches of) persistence diagrams.
+
+    :param pdiagset: a set of persistence diagrams that we want to quantize.
+    :type pdiagset: list of n x 2 numpy arrays
+    :param k: number of centroids. Default is ``2``. A naive heuristic is used to initialize the codebook.
+              Not used if an initial codebook ``c0`` has been provided.
+    :type k: ``int``
+    :param init: If provided, overwrite the provided ``k`` (which is now ``c0.shape[0]``), and is used as an
+                initialization for the algorithm. Default is ``None``.
+    :type init: (n x 2) ``numpy.array``, or ``None``.
+    :param batch_size: Size of batches used during the online exploration of the ``pdiagset``.
+                        Default is ``1``.
```

Comment:
As a user, should I stick to the default value of 1? If I already have all the diagrams, I may think that I don't need an online algorithm, which is for when data appears progressively, and consider using one huge batch under the impression that it disables the "online" stuff and gets the best result.